src/Common/src/Microsoft/Win32/SafeHandles/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Allows limited reuse of heap buffers to improve memory pressure. This cache does not ensure
+    /// that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
+    {
+        private readonly ulong _minSize;
+        private readonly ulong _maxSize;
+
+        private LowLevelConcurrentStack<SafeHeapHandle> handleCache = new LowLevelConcurrentStack<SafeHeapHandle>();
```

Comment:
> Taking that out

Who's suggesting taking that out?  The scheme we used to have and that I'm talking about here reuses existing space.

> even at two legacy max paths that is 1MB

I don't understand the math here.  I assume by legacy max path you mean 260 chars?  Which would be 520 bytes.  1MB of those would mean ~2,000 buffers.  How are you calculating 1MB?

> it negatively impacted the measurements I took

You could certainly come up with benchmarks where it would lose.  You can also come up with benchmarks where it significantly wins.  I'm suggesting that the cases where it wins are much more common than the cases where it loses.  Can you please share the benchmark you used?

> I think it makes more sense to add it in once we've found a real measurement that shows how it improves a specific scenario

We currently use scheme A.  You're proposing jumping from scheme A to scheme B and then saying we need measurements to justify going back to scheme A.  I don't get it. Can you share the commits you used to do the measurements that provide the evidence you're talking about?

> this all about optimizing scratch space. We don't need to zero, we need to be able to grow quickly

I didn't say anything about zeroing.

> I took the current behavior and dug into how to make it better using the original patterns to inform my attack on this. I'm not arbitrarily making changes- everything has many hours of investigation behind it.

I'm not saying you're making arbitrary changes.  I'm saying your'e making a whole bag of changes all together, and in doing so it's not clear that each individual change is a net win.  I'm saying that maintaining a global, never-shrinking pool of buffers can be quite wasteful.  I'm saying that we've significantly increased the amount of unsafe code being used, and I'm concerned we're introduced subtle bugs that will be difficult to catch.  I'm saying that these changes introduce additional allocations that are difficult to quantify as part of microbenchmarks but that can have an effect under real-world load.